akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.stream.ActorMaterializer
+import akka.http.scaladsl.settings.ServerSettings
+import com.typesafe.config.{ Config, ConfigFactory }
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, ExecutionContext, Future }
+import scala.io.StdIn
+import scala.util.{ Failure, Success }
+
+/**
+ * Bootstrap trait. It helps booting up an akka-http by only defining the desired routes.
+ * It offers additional hooks to modify the default behavior
+ */
+trait HttpApp {
+
+  def main(args: Array[String]) = {
+    val config = createConfig()
+    implicit val system = createActorSystem(config)
+    implicit val materializer = ActorMaterializer()
+    implicit val executionContext = system.dispatcher
+    postStartActorSystem(system)
+
+    val host = config.getString("akka.http.server.host")
+    val port = config.getInt("akka.http.server.port")
```

Comment:
My personal preference would be for this not be a JVM-`main(args: Array[String])` method, but instead some kind of 
```scala
def startServer(host: String, port: Int, settings: ServerSettings, system: ActorSystem = null)
```

Having it take in the `system` as an optional argument-with-default would also make this more useful for people who want to start up servers in an existing Akka project. Also, I often like converting config into to type-safe settings objects as soon as possible rather than waiting for some library to try (and possibly fail) on my behalf.

Letting people deal with the `ActorSystem` externally will also remove the need for the hooks related to it, since they would already have the option of instantiating their own system and running whatever code they wish before or after.

Generally I think people will want more flexibility when starting the server: e.g. setting the port based on some dynamically-loaded configuration, or running multiple servers on different ports in the same JVM. Even for my current tiny project, i have a bunch of custom initialization I need to do before starting the server, and I assume it'll only grow as projects get larger. There is no inherent reason why starting the server should be tied to starting the JVM process, and we can provide the helper-ness of dealign with all the boilerplate while still providing the flexibility to start servers where-ever someone wants to